rclcpp/test/rclcpp/test_graph_listener.cpp

```diff
@@ -0,0 +1,309 @@
+};
+
+/* Run base class init/shutdown */
+TEST_F(TestGraphListener, construction_and_destruction) {
+  void();
+}
+
+// Required for mocking_utils below
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, ==)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, !=)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, <)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcl_guard_condition_options_t, >)
+
+/* Error creating a new graph listener */
+TEST_F(TestGraphListener, error_construct_graph_listener) {
+  using rclcpp::contexts::get_global_default_context;
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_guard_condition_init, RCL_RET_ERROR);
+
+  using rclcpp::contexts::get_global_default_context;
+  RCLCPP_EXPECT_THROW_EQ(
+  {
+    auto graph_listener_error =
+    std::make_shared<rclcpp::graph_listener::GraphListener>(get_global_default_context());
+  }, std::runtime_error("failed to create interrupt guard condition: error not set"));
+}
+
+// Required for mocking_utils below
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, ==)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, !=)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, <)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, >)
+
+/* Errors that occur when initializing the graph_listener */
+TEST_F(TestGraphListener, error_start_graph_listener) {
+  {
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_wait_set_init, RCL_RET_ERROR);
+    RCLCPP_EXPECT_THROW_EQ(
+    {
+      graph_listener()->start_if_not_started();
+    }, std::runtime_error("failed to initialize wait set: error not set"));
+  }
+  {
+    auto context_to_destroy = std::make_shared<rclcpp::contexts::DefaultContext>();
+    context_to_destroy->init(0, nullptr);
+    auto graph_listener_error =
+      std::make_shared<rclcpp::graph_listener::GraphListener>(context_to_destroy);
+    // rclcpp::shutdown(context_to_destroy); Using this causes test to fail
+    context_to_destroy.reset();
+    RCLCPP_EXPECT_THROW_EQ(
+    {
+      graph_listener_error->start_if_not_started();
+    }, std::runtime_error("parent context was destroyed"));
+  }
+  {
+    EXPECT_NO_THROW(graph_listener()->shutdown());
+    RCLCPP_EXPECT_THROW_EQ(
+    {
+      graph_listener()->start_if_not_started();
+    }, std::runtime_error(shutdown_error_str));
+  }
+}
+
+class TestGraphListenerProtectedMethods : public rclcpp::graph_listener::GraphListener
+{
+public:
+  explicit TestGraphListenerProtectedMethods(std::shared_ptr<rclcpp::Context> parent_context)
+  : rclcpp::graph_listener::GraphListener{parent_context}
+  {}
+
+  void run_protected()
+  {
+    this->run();
+  }
+};
+
+// Tricky: patch on return the start method, to prepare the
+// loop thread to be run, but leave early with the failure thrown
+// Then, run the loop method to test it
+#define MOCK_WAIT_SET_INIT_ON_THREAD() \
```

Comment:
This is cool. Does this need to be a macro, or could you just create a member function for your test fixture? Also, it might be good to have an additional test that just checks this case without additional mocks

Reply:
I used a macro because the code is mocked, and AFAIK mocking inside a function will erase the mock after leaving the scope of the function.

As for the the mocked function, is tested on its own as the first test of `error_start_graph_listener`, do you mean a test on its own for the macro itself?

Reply:
Oh, you are correct about the function scope issue. But that raises another question, why doesn't `mock_wait_set_init` fall out of scope in this do-while loop?

You might just have to take the inject_on_error call and duplicate in each block below. The expected exception can just be checked the first time

Reply:
The purpose of the mocking function is to initialize the graph_listener variables without actually starting it. I just realized the mock goes out of scope, but the graph_listener stays in the desired state. So I think I can change it to be a member function of the test fixture.

Reply:
Changed this with: 06b3645